pkg/util/clusteroperator/status.go

```diff
@@ -1,65 +1,223 @@
+	reasonOperatorDisabled   = "OperatorDisabledByAdmin"
+	msgOperatorDisabled      = "Credential minting is disabled by cluster admin"
+)
 
-	newConditions := []configv1.ClusterOperatorStatusCondition{}
+// StatusHandler produces conditions and related objects to be reflected
+// in the cloud-credential-operator ClusterOperatorStatus
+type StatusHandler interface {
+	GetConditions(logger log.FieldLogger) ([]configv1.ClusterOperatorStatusCondition, error)
+	GetRelatedObjects(logger log.FieldLogger) ([]configv1.ObjectReference, error)
+	Name() string
+}
 
-	found := false
-	for _, c := range oldConditions {
-		if condition.Type == c.Type {
-			if condition.Status == c.Status &&
-				condition.Reason == c.Reason &&
-				condition.Message == c.Message {
-				return oldConditions
-			}
+var (
+	statusHandlers = []StatusHandler{}
+)
 
-			found = true
-			newConditions = append(newConditions, *condition)
-		} else {
-			newConditions = append(newConditions, c)
+// AddStatusHandler registers a StatusHandler that will be called whenever
+// an update to the ClusterOperator status is requested
+func AddStatusHandler(handler StatusHandler) {
+	statusHandlers = append(statusHandlers, handler)
+}
+
+// SyncStatus syncs the ClusterOperator status by calling all registered
+// status handlers for conditions and related objects
+func SyncStatus(client client.Client, logger log.FieldLogger) error {
+	logger.Debug("syncing cluster operator status")
+
+	co := &configv1.ClusterOperator{ObjectMeta: metav1.ObjectMeta{Name: cloudCredClusterOperator}}
+	err := client.Get(context.TODO(), types.NamespacedName{Name: co.Name}, co)
+	isNotFound := errors.IsNotFound(err)
+	if err != nil && !isNotFound {
+		return fmt.Errorf("failed to get clusteroperator %s: %v", co.Name, err)
+	}
+
+	oldConditions := co.Status.Conditions
+	oldVersions := co.Status.Versions
+	oldRelatedObjects := co.Status.RelatedObjects
+
+	conditions := []configv1.ClusterOperatorStatusCondition{}
+	relatedObjects := []configv1.ObjectReference{}
+	for _, handler := range statusHandlers {
+		handlerConditions, err := handler.GetConditions(logger)
+		if err != nil {
+			logger.Errorf("failed to get conditions from handler %s", handler.Name())
+			continue
+		}
+		conditions = mergeConditions(conditions, handlerConditions, logger, handler.Name())
+		handlerRelatedObjects, err := handler.GetRelatedObjects(logger)
+		if err != nil {
+			logger.Errorf("failed to get related objects from handler %s", handler.Name())
+			continue
 		}
+		relatedObjects = append(relatedObjects, handlerRelatedObjects...)
+	}
+
+	// sets defaults for condition not set by any handler
+	conditions = defaultUnsetConditions(conditions)
+
+	// at this point we known all condition types exist in conditions
+
+	co.Status.Conditions = conditions
+	co.Status.RelatedObjects = relatedObjects
+	co.Status.Versions = computeClusterOperatorVersions()
+
+	operatorIsDisabled, err := utils.IsOperatorDisabled(client, logger)
```

Comment:
the reason I put it here is so that i don't have to do the little dance checking to see if the condition exists and if it doesn't create/append it, etc.

though you did make me think of an issue with this.  we shouldn't just blindly overwrite the reason if the condition was set `False` by a handler.  i'll fix that.

Reply:
plus it isn't changing the `Status` on the condition, just the `Reason` and `Message`.